akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -357,25 +362,57 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
         }
       })
 
+      // -----------------------------------------------------------------
       // FIXME: What if user handler doesn't pull in new substreams? Should we reject them
-      //        after a while or buffer only a limited amount? We should also be able to
-      //        keep the buffer limited to the number of concurrent streams as negotiated
-      //        with the other side.
+      //        after a while or buffer only a limited amount?
       val bufferedSubStreamOutput = new BufferedOutlet[Http2SubStream](substreamOut)
       override def dispatchSubstream(initialHeaders: ParsedHeadersFrame, data: Source[Any, Any], correlationAttributes: Map[AttributeKey[_], _]): Unit =
         bufferedSubStreamOutput.push(Http2SubStream(initialHeaders, data, correlationAttributes))
 
+      // -----------------------------------------------------------------
+      private var completing = false
+      private def onTryComplete(): Unit = {
+        completing = true
+        if (safeToComplete) complete()
+      }
+      override def onAllStreamsClosed(): Unit = if (safeToComplete) complete()
+      override def onMultiplexerIdle(): Unit = if (safeToComplete) complete()
+
+      private def safeToComplete = {
+        if (isServer) completing && multiplexer.isIdle && activeStreamCount() == 0
```

Comment:
Instead of waiting for `multiplexer.isIdle` before declaring `safeToComplete` and calling `complete()`, I wonder if we could call `complete()` immediately but replace `complete(frameOut)` by a (to be implemented) `multiplexer.complete()`, and make the multiplexer responsible for eventually completing `frameOut` - similar to what we did for `bufferedSubStreamOutput`.